Lilu/Headers/kern_util.hpp

```diff
@@ -944,4 +945,285 @@ class evector_base {
+	 */
+	static inline void safeDestory(CircularBuffer<T> *&buffer) {
+		if (buffer != nullptr) {
+			destory(buffer);
+			buffer = nullptr;
+		}
+	}
+	
+	/**
+	 *  Check whether the circular buffer is empty
+	 *
+	 *  @return `true` if the buffer is empty, `false` otherwise.
+	 */
+	inline bool isEmpty() {
+		IORecursiveLockLock(lock);
+		bool retVal = (count == 0) && (indexr == indexw);
+		IORecursiveLockUnlock(lock);
+		return retVal;
+	}
+	
+	/**
+	 *  Check whether the circular buffer is full
+	 *
+	 *  @return `true` if the buffer is full, `false` otherwise.
+	 */
+	inline bool isFull() {
+		IORecursiveLockLock(lock);
+		bool retVal = (count == size) && (indexr == indexw);
+		IORecursiveLockUnlock(lock);
+		return retVal;
+	}
+	
+	/**
+	 *  Get the number of elements in the circular buffer
+	 *
+	 *  @return The current number of elements in the buffer.
+	 */
+	inline IOItemCount getCount() {
+		IORecursiveLockLock(lock);
+		IOItemCount retVal = count;
+		IORecursiveLockUnlock(lock);
+		return retVal;
+	}
+	
+	/**
+	 *  Write the given element to the circular buffer
+	 *
+	 *  @param element The element to write
+	 *  @return `true` on success, `false` if the buffer is full.
+	 */
+	inline bool write(const T &element) {
+		if (isFull())
+			return false;
+		
+		IORecursiveLockLock(lock);
+		storage[indexw] = element;
+		indexw += 1;
+		indexw %= size;
+		count += 1;
+		IORecursiveLockUnlock(lock);
+		return true;
+	}
+	
+	/**
+	 *  Read the next element from the circular buffer
+	 *
+	 *  @param element The element read from the buffer
+	 *  @return `true` on success, `false` if the buffer is empty.
+	 */
+	inline bool read(T& element) {
+		if (isEmpty())
+			return false;
+		
+		IORecursiveLockLock(lock);
+		element = storage[indexr];
+		indexr += 1;
+		indexr %= size;
+		count -= 1;
+		IORecursiveLockUnlock(lock);
+		return true;
+	}
+};
+
+/**
+ *  Wrap an object that is not an instance of OSObject
+ */
+class OSObjectWrapper: public OSObject {
+	/**
+	 *  Constructors & Destructors
+	 */
+	OSDeclareDefaultStructors(OSObjectWrapper);
+	
+	using super = OSObject;
+	
+	/**
+	 *  Wrapped object
+	 */
+	void *object;
+	
+public:
+	/**
+	 *  Initialize the wrapper with the given object
+	 *
+	 *  @param object The wrapped object that is not an `OSObject`
+	 *  @return `true` on success, `false` otherwise.
+	 */
+	EXPORT bool init(void *object);
+	
+	/**
+	 *  Reinterpret the wrapped object as the given type
+	 *
+	 *  @return The wrapped object of the given type.
+	 */
+	template <typename T>
+	T *reinterpretWrappedObjectAs() {
+		return reinterpret_cast<T*>(object);
+	}
+	
+	/**
+	 *  Create a wrapper for the given object that is not an `OSObject`
+	 *
+	 *  @param object A non-null object
+	 *  @return A non-null wrapper on success, `nullptr` otherwise.
+	 *  @warning The caller is responsbile for managing the lifecycle of the given object.
+	 */
+	EXPORT static OSObjectWrapper *of(void *object);
```

Comment:
In my opinion, `with`, fits better.

Reply:
How do we free the object? Should we pass an optional method as a second param here?

Reply:
No, `OSObjectWrapper` does not manage the lifecycle of the wrapped object, so we don't need to specify the deleter.

Reply:
Not sure I like it, but let's go with it for now.